nat/service_ics.go

```diff
@@ -0,0 +1,147 @@
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.EnableSharing(0)`)
+
+	return err
+}
+
+func (nat *serviceICS) Add(rule RuleForwarding) error {
+	// Enable internet connection sharing for the local interface.
+	// TODO detect interface name from the rule.
+	// TODO firewall rule configuration should be added here for new connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+	$netShare = New-Object -ComObject HNetCfg.HNetShare;
+	$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+	$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+	$config.EnableSharing(1)`)
+
+	return err
+}
+
+func (nat *serviceICS) Del(rule RuleForwarding) error {
+	// Disable internet connection sharing for the local interface.
+	// TODO detect interace name from the rule.
+	// TODO firewall rule configuration should be added here for cleaning up unused connections.
+	ifaceName := "myst1"
+
+	_, err := powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func (nat *serviceICS) Disable() error {
+	// TODO stop internet connection sharing on all enabled interfaces (nat.ifaces).
+	// TODO we should cleanup as much as possible, not failing after errors.
+	_, err := powerShell("Set-Service -Name RemoteAccess -StartupType " + nat.remoteAccessStatus)
+	if err != nil {
+		return err
+	}
+
+	ifaceName, err := getPublicInterfaceName()
+	if err != nil {
+		return err
+	}
+
+	// Disable internet connection sharing for the public interface.
+	_, err = powerShell(`regsvr32 /s hnetcfg.dll;
+		$netShare = New-Object -ComObject HNetCfg.HNetShare;
+		$c = $netShare.EnumEveryConnection |? { $netShare.NetConnectionProps.Invoke($_).Name -eq "` + ifaceName + `" };
+		$config = $netShare.INetSharingConfigurationForINetConnection.Invoke($c);
+		$config.DisableSharing()`)
+
+	return err
+}
+
+func getPublicInterfaceName() (string, error) {
+	out, err := powerShell(`Get-WmiObject -Class Win32_IP4RouteTable | where { $_.destination -eq '0.0.0.0' -and $_.mask -eq '0.0.0.0'} | foreach { $_.InterfaceIndex }`)
+	if err != nil {
+		return "", err
+	}
+
+	ifaceID, err := strconv.Atoi(strings.TrimSpace(string(out)))
+	if err != nil {
+		return "", err
+
+	}
+
+	ifaces, err := net.Interfaces()
+	if err != nil {
+		return "", err
+	}
+
+	for _, iface := range ifaces {
+		if iface.Index == ifaceID {
+			return iface.Name, nil
+		}
+	}
+
+	return "", errors.New("interface not found")
+}
+
+func powerShell(cmd string) ([]byte, error) {
+	out, err := exec.Command("powershell", "-Command", cmd).CombinedOutput()
+	return out, errors.Wrap(err, string(out))
```

Comment:
You wrap perfectly legit Command output as an error all the time. Maybe simple printout would do for normal output, on error, you can warp it like you did.

Reply:
If the `err` is `nil` output will be ignored, it adds an output to the real errors only, to add some context and avoid errors like `exit status 1` with no context.